cou/steps/hypervisor.py

```diff
@@ -0,0 +1,152 @@
+# Copyright 2024 Canonical Limited
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Hypervisor planner."""
+import logging
+from dataclasses import dataclass
+
+from cou.apps.base import ApplicationUnit, OpenStackApplication
+from cou.steps import (
+    HypervisorUpgradePlan,
+    PostUpgradeStep,
+    PreUpgradeStep,
+    UpgradePlan,
+)
+from cou.utils.openstack import OpenStackRelease
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass(frozen=True)
+class Hypervisor:
+    """Hypervisor containing units for multiple applications."""
+
+    name: str
+    units: list[ApplicationUnit]
+
+
+@dataclass(frozen=True)
+class AZ:
+    """Availability Zone hypervisors in the az."""
+
+    name: str
+    hypervisors: list[Hypervisor]
+
+
+def verify_apps(apps: list[OpenStackApplication]) -> None:
+    """Verify OpenStack applications.
+
+    This function will verify that all machines contain nova-compute application.
+
+    :param apps: list of OpenStack applications
+    :type apps: list[OpenStackApplication]
+    """
+    raise NotImplementedError
+
+
+class HypervisorUpgradePlanner:
+    """Planer for all hypervisor updates.
+
+    This planner is meant to be used to upgrade machines contains the nova-compute application.
+    """
+
+    def __init__(self, apps: list[OpenStackApplication]) -> None:
+        """Initialize the Hypervisor class.
+
+        :param apps: list of OpenStack applications
+        :type apps: list[OpenStackApplication]
+        """
+        verify_apps(apps)
+        self._apps = apps
+
+    @property
+    def apps(self) -> list[OpenStackApplication]:
+        """Return list of apps in hypervisor class."""
+        return self._apps
+
+    @property
+    def azs(self) -> list[AZ]:
```

Comment:
It's unclear to me how the AZs are determined through application. But since this is only the base structure, we can leave it as "magic" behind the scene

Reply:
yes, this will be more tricky than just for loop, but I did not want to focus on that here. Next PR